sql/src/main/java/org/apache/druid/sql/SqlLifecycle.java

```diff
@@ -0,0 +1,361 @@
+      return sqlQueryId();
+    }
+  }
+
+  public String initialize(String sql, Map<String, Object> queryContext)
+  {
+    synchronized (lock) {
+      transition(State.NEW, State.INITIALIZED);
+      this.sql = sql;
+      this.queryContext = contextWithSqlId(queryContext);
+      return sqlQueryId();
+    }
+  }
+
+  private Map<String, Object> contextWithSqlId(Map<String, Object> queryContext)
+  {
+    Map<String, Object> newContext = Maps.newHashMap();
+    if (queryContext != null) {
+      newContext.putAll(queryContext);
+    }
+    if (!newContext.containsKey(PlannerContext.CTX_SQL_QUERY_ID)) {
+      newContext.put(PlannerContext.CTX_SQL_QUERY_ID, UUID.randomUUID().toString());
+    }
+    return newContext;
+  }
+
+  private String sqlQueryId()
+  {
+    return (String) this.queryContext.get(PlannerContext.CTX_SQL_QUERY_ID);
+  }
+
+  public PlannerContext plan(AuthenticationResult authenticationResult)
+      throws ValidationException, RelConversionException, SqlParseException
+  {
+    synchronized (lock) {
+      transition(State.INITIALIZED, State.PLANED);
+      try (DruidPlanner planner = plannerFactory.createPlanner(queryContext, authenticationResult)) {
+        this.plannerContext = planner.getPlannerContext();
+        this.plannerResult = planner.plan(sql);
+      }
+      return plannerContext;
+    }
+  }
+
+  public PlannerContext plan(HttpServletRequest req)
+      throws SqlParseException, RelConversionException, ValidationException
+  {
+    synchronized (lock) {
+      this.req = req;
+      return plan(AuthorizationUtils.authenticationResultFromRequest(req));
+    }
+  }
+
+  public RelDataType rowType()
+  {
+    synchronized (lock) {
+      Preconditions.checkState(plannerResult != null,
+                               "must be call after sql has been planned");
+      return plannerResult.rowType();
+    }
+  }
+
+  public Access authorize()
+  {
+    synchronized (lock) {
+      transition(State.PLANED, State.AUTHORIZING);
+
+      if (req != null) {
+        return doAuthorize(
+            plannerContext.getAuthenticationResult(),
+            AuthorizationUtils.authorizeAllResourceActions(
+                req,
+                Iterables.transform(
+                    plannerResult.datasourceNames(),
+                    AuthorizationUtils.DATASOURCE_READ_RA_GENERATOR
+                ),
+                plannerFactory.getAuthorizerMapper()
+            )
+        );
+      }
+
+      return doAuthorize(
+          plannerContext.getAuthenticationResult(),
+          AuthorizationUtils.authorizeAllResourceActions(
+              plannerContext.getAuthenticationResult(),
+              Iterables.transform(plannerResult.datasourceNames(), AuthorizationUtils.DATASOURCE_READ_RA_GENERATOR),
+              plannerFactory.getAuthorizerMapper()
+          )
+      );
+    }
+  }
+
+  private Access doAuthorize(final AuthenticationResult authenticationResult, final Access authorizationResult)
```

Comment:
The `authenticationResult` is unused here

Reply:
will remove later